video/audio_sample.h

```diff
@@ -8,61 +8,51 @@
 #include "audio_channel.h"
 #include "buffer_stream.h"
 
-struct audio_sample {
-	audio_sample(std::vector<std::shared_ptr<BufferStream>>& streams, uint8_t format) : blocks(streams), format(format), index(0), blockIndex(0) {}
-	~audio_sample();
-	int8_t getSample();
-	void rewind() {
-		index = 0;
-		blockIndex = 0;
-	}
-	void checkIndexes() {
-		if (blockIndex >= blocks.size()) {
-			blockIndex = 0;
-			index = 0;
-		}
-		if (index >= blocks[blockIndex]->size()) {
-			index = 0;
-		}
-	}
-	uint32_t getDuration() {
-		uint32_t duration = 0;
-		for (auto block : blocks) {
-			duration += block->size();
-		}
-		return duration / 16;
-	}
+struct AudioSample {
+	AudioSample(std::vector<std::shared_ptr<BufferStream>>& streams, uint8_t format, uint32_t sampleRate = AUDIO_DEFAULT_SAMPLE_RATE, uint16_t frequency = 0) :
+		blocks(streams), format(format), sampleRate(sampleRate), baseFrequency(frequency) {}
+	~AudioSample();
+
+	int8_t getSample(uint32_t & index, uint32_t & blockIndex);
+	void seekTo(uint32_t position, uint32_t & index, uint32_t & blockIndex, int32_t & repeatCount);
+	uint32_t getSize();
+	uint32_t getDuration();
+
 	std::vector<std::shared_ptr<BufferStream>>& blocks;
-	uint8_t			format;			// Format of the sample data
-	uint32_t		index;			// Current index inside the current sample block
-	uint32_t		blockIndex;		// Current index into the sample data blocks
```

Comment:
AudioSample objects now no longer keep track of indexes into the sample data themselves

it was flawed to do so - as it meant that if an individual sample was played back on multiple channels it would just play back faster.  oops!  😀 

this is now tracked instead by the sample generator.  functions for managing stepping thru the sample however remain in here, as the AudioSample object is what understands how the data is structured